stdlib/native.rb

```diff
@@ -160,6 +160,44 @@ def Array(object, *args, &block)
       }
     }
   end
+  
+  # allows to pass a proc that runs in context of `this`
+  #
+  # @example
+  #   options = {
+  #     persist: false,
+  #     render: function { |item, escape|
+  #       '<div>' + escape.call(item.name) + '</div>'
+  #     }
+  #   }
+  #
+  def function *args, &block
+    args.any? && raise(ArgumentError, '`function` does not accept arguments')
+    block || raise(ArgumentError, 'block required')
+    proc do |*a| # no way to expect a block here as this is always called from JavaScript
+      a.map! {|x| Native(`x`)}
+      instance = Native(`this`)
+      %x{
+        // if window is current scope, run the block in the scope it was defined
+        if(this.toString() === '[object Window]') {
+          return block.apply(self, a);
+        }
+        // otherwise run the block in the current scope
+        return (function() {
+          var s = block.$$s;
```

Comment:
i have very vague ideas why this is done, i just stolen it from stdlib's `instance_exec` :)
https://github.com/opal/opal/blob/dc8f5fa5545ccb46b48753e1ea6e471591346dc6/opal/corelib/basic_object.rb#L58